### PR TITLE
quad: Import sympy only when necessary

### LINKED ISSUE #449: Startup time is slow
I timed each of
```
import numpy
import quantecon
```

The numpy import took 0.288s. The quantecon import took 5.007s.
From `python -X importtime` I found the most time consuming imports are (not exhaustive because the importtime output is too big to parse by eye):
(unit: in ms)
1. 636 quantecon.arma (mainly scipy.signal)
2. 2820 quantecon.random
3. 1121 quantecon.quad (mainly sympy)
4. 520 quantecon.game_theory
5. 507 numba

Should quantecon.random and quantecon.quad not be imported at the main namespace?

### PR DESCRIPTION
**Pull request by rht**
This ~partially addresses~ fixes #449. The `import quantecon` time on my machine decreases from ~5.4s to ~4.6s~ to ~1s.

### CODE REVIEW

**Comment by rht** [comment]
With the second commit, it goes from ~4.6s to ~2s.

**Comment by rht** [comment]
`nose-timer` stat:
```
[success] 14.59% quantecon.tests.test_gridtools.test_performance_C: 31.6635s
[success] 10.93% quantecon.tests.test_gridtools.test_performance_F: 23.7086s
[success] 3.17% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_bohachevsky: 6.8868s
[success] 2.98% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_zakharov: 6.4733s
[success] 2.66% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_sum_squared: 5.7616s
[success] 2.50% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_styblinski_tang: 5.4286s
[success] 2.42% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_perm_function: 5.2586s
[success] 2.42% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_easom: 5.2527s
[success] 2.35% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_rotated_hyper_ellipsoid: 5.1085s
[success] 2.28% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_goldstein_price: 4.9372s
[success] 2.26% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_powell: 4.9080s
[success] 2.20% quantecon.util.tests.test_numba.TestNumbaLinalgSolve.test_b_1dim: 4.7764s
[success] 2.18% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_rosenbrock: 4.7321s
[success] 2.16% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_corner_sol: 4.6921s
[success] 2.13% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_colville: 4.6129s
[success] 2.12% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_mccormick: 4.6027s
[success] 2.10% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_booth: 4.5612s
[success] 2.09% quantecon.optimize.tests.test_nelder_mead.TestMaximization.test_discontinuous: 4.5425s
[success] 1.75% quantecon.optimize.tests.test_nelder_mead.test_invalid_σ: 3.7913s
[success] 1.74% quantecon.optimize.tests.test_nelder_mead.test_invalid_ρχ: 3.7754s
[success] 1.71% quantecon.optimize.tests.test_nelder_mead.test_invalid_γ: 3.7168s
[success] 1.63% quantecon.optimize.tests.test_nelder_mead.test_invalid_ρ: 3.5275s
[success] 1.60% quantecon.optimize.tests.test_nelder_mead.test_invalid_χ: 3.4731s
[success] 1.34% quantecon.util.tests.test_numba.TestNumbaLinalgSolve.test_b_2dim: 2.9008s
[success] 1.26% test_lemke_howson.TestLemkeHowson.test_lemke_howson: 2.7230s
[success] 1.20% test_repeated_game.TestAS.test_abreu_sannikov: 2.5982s
[success] 1.15% test_random.TestRandomDiscreteDP.test_equal_beta: 2.5023s
[success] 1.15% quantecon.tests.test_inequality.test_lorenz_curve: 2.4939s
[success] 1.07% test_mclennan_tourky.TestEpsilonNash.test_epsilon_nash_with_full_output: 2.3290s
[success] 0.94% test_random.test_random_stochastic_matrix_dense_vs_sparse: 2.0447s
[success] 0.86% quantecon.tests.test_compute_fp.TestComputeFPContraction.test_num_iter_large: 1.8700s
[success] 0.83% quantecon.optimize.tests.test_nelder_mead.test_invalid_bounds_1: 1.7947s
[success] 0.78% test_support_enumeration.TestSupportEnumeration.test_no_error_skew_sym: 1.6925s
[success] 0.76% quantecon.tests.test_lss.TestLinearStateSpace.test_replicate: 1.6590s
[success] 0.69% test_support_enumeration.TestSupportEnumeration.test_support_enumeration: 1.5076s
[success] 0.67% test_utilities.test_sample_without_replacement_shape: 1.4527s
[success] 0.66% test_utilities.TestProbvec.test_parallel_cpu: 1.4344s
[success] 0.64% quantecon.tests.test_inequality.test_gini_coeff: 1.3894s
[success] 0.58% test_random.test_random_markov_chain_dense: 1.2546s
[success] 0.58% test_normal_form_game.test_player_corner_cases: 1.2525s
[success] 0.53% test_random.TestRandomDiscreteDP.test_shape: 1.1552s
[success] 0.53% test_utilities.TestProbvec.test_shape: 1.1430s
[success] 0.49% test_vertex_enumeration.TestVertexEnumeration.test_vertex_enumeration: 1.0575s
[success] 0.48% test_random.TestRandomDiscreteDP.test_nonzero: 1.0440s
[success] 0.48% quantecon.tests.test_compute_fp.TestComputeFPContraction.test_2d_input: 1.0360s
[success] 0.48% test_random.TestRandomDiscreteDP.test_equal_reward: 1.0338s
[success] 0.48% test_random.TestRandomDiscreteDP.test_equal_probability: 1.0319s
[success] 0.46% quantecon.util.tests.test_timing.TestTicTacToc.test_loop: 1.0033s
[success] 0.44% test_bimatrix_generators.TestTournamentGame.test_payoff_values: 0.9591s
[success] 0.43% test_ddp.test_ddp_beta_0: 0.9365s
[success] 0.36% quantecon.optimize.tests.test_root_finding.test_brentq_basic: 0.7826s
[success] 0.36% quantecon.optimize.tests.test_scalar_max.test_brent_max: 0.7812s
[success] 0.36% test_ddp.TestDiscreteDP.test_modified_policy_iteration: 0.7807s
[success] 0.35% quantecon.optimize.tests.test_root_finding.test_newton_basic: 0.7537s
[success] 0.33% quantecon.optimize.tests.test_scalar_max.test_invalid_a_brent_max: 0.7132s
[success] 0.33% test_normal_form_game.test_best_response_2p: 0.7081s
[success] 0.33% test_random.test_random_stochastic_matrix_sparse: 0.7054s
[success] 0.32% test_random.test_random_stochastic_matrix_dense: 0.6928s
[success] 0.30% test_random.test_random_markov_chain_sparse: 0.6608s
[success] 0.28% quantecon.optimize.tests.test_scalar_max.test_invalid_a_b_brent_max: 0.6117s
[success] 0.28% quantecon.optimize.tests.test_root_finding.test_newton_hard: 0.6071s
[success] 0.27% test_bimatrix_generators.TestBlottoGame.test_constant_diagonal: 0.5965s
[success] 0.26% quantecon.optimize.tests.test_scalar_max.test_invalid_b_brent_max: 0.5594s
[success] 0.22% quantecon.optimize.tests.test_root_finding.test_bisect_basic: 0.4668s
[success] 0.21% quantecon.optimize.tests.test_root_finding.test_halley_hard: 0.4654s
[success] 0.20% quantecon.optimize.tests.test_root_finding.test_halley_basic: 0.4394s
```

**Comment by rht** [comment]
`test_nelder_mead` alone took 46.45% of the entire time. Since Travis can have 4 concurrent jobs, it would be possible to split the test into the cartesian product of [linux, darwin] and [`test_nelder_mead`, the rest].

**Comment by oyamad** [comment]
Do we really need these tests

> [success] 14.59% quantecon.tests.test_gridtools.test_performance_C: 31.6635s
> [success] 10.93% quantecon.tests.test_gridtools.test_performance_F: 23.7086s

?

Ref #109

**Comment by oyamad** [comment]
Adding `target='parallel'` to the `_probvec_*` functions reduced the import time from 1.922s to 1.003s on the *second time* import on my machine.

**Comment by rht** [comment]
OK, I removed the guvectorize commit. I suppose I should have timed the hot start that has the numba caches ready.

**Comment by rht** [comment]
> Adding `target='parallel'` to the `_probvec_*` functions reduced the import time from 1.922s to 1.003s on the _second time_ import on my machine.

@oyamad did you mean `cache=True`?

**Comment by rht** [comment]
On `quantecon.arma`, either:
1. not put it on the top-level name space (`scipy.signal` is not at the `scipy` namespace either, https://github.com/scipy/scipy/blob/master/scipy/__init__.py)
2. lazy-import the `scipy.signal` at the class initialization

**Comment by oyamad** [comment]
@rht Yes, `cache=True`, sorry missed copy-paste...

**Comment by oyamad** [comment]
+1 for 2. lazy import of `scipy.signal`.

File: quantecon/random/utilities.py [hunk 1]
```diff
@@ -89,10 +89,10 @@ def _probvec(r, out):
     out[n] = 1 - r[n-1]
 
 _probvec_parallel = guvectorize(
-    ['(f8[:], f8[:])'], '(n), (k)', nopython=True, target='parallel'
+    ['(f8[:], f8[:])'], '(n), (k)', nopython=True, target='parallel', cache=True,
```
Comment (oyamad) [inline quantecon/random/utilities.py]: You don't need the comma at the end.
And line too long?

**Comment by rht** [comment]
OK, the import time is now at ~1s on my machine,
2018-12-12-110234_1863x900_scrot
Other than the `numba` import, the rest consists of the cached generated_jit and guvectorize, which I suppose can't be optimized further.

**Comment by mmcky** [comment]
thanks @oyamad and @rht this is really great work! That's a pretty neat tool.

**Comment by mmcky** [comment]
I think selective imports works well -- when done like this. Is there any PEP guidance to this?

**Comment by rht** [comment]
The only article I found documenting various options to do lazy import: https://snarky.ca/lazy-importing-in-python-3-7/ (note: the author wrote `importlib.util.LazyLoader` for the stdlib, https://github.com/python/cpython/commit/a04dbe4fe7c2bacfb48b338ed4248a58c540beea). From the article, I suppose the demand for the functionality was substantial enough to have it formalized in `importlib` instead of resorting the local imports method I have used here.

**Comment by rht** [comment]
See also: https://github.com/numpy/numpy/issues/11457.

**Comment by mmcky** [comment]
thanks @rht - this is a great improvement! Also thanks for those links. Lazy importing makes a lot of sense.

**Pull request merged**